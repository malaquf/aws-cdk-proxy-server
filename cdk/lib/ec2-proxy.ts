import {Fn, Stack, StackProps} from "aws-cdk-lib";
import {Construct} from "constructs";
import {
    AmazonLinuxImage,
    CfnInstance,
    InstanceClass,
    InstanceSize,
    InstanceType,
    SecurityGroup,
    UserData,
    Vpc
} from "aws-cdk-lib/aws-ec2";
import {CfnInstanceProfile, ManagedPolicy, Role, ServicePrincipal} from "aws-cdk-lib/aws-iam";
import { StringParameter } from "aws-cdk-lib/aws-ssm";

export class Ec2Stack extends Stack {

    constructor(scope: Construct, id: string, props?: StackProps) {
        super(scope, id, props);

        const vpc = Vpc.fromLookup(this, 'ImportedVpc', {
            vpcId: StringParameter.valueFromLookup(this, '/VpcId'),
        });

        const ec2SecurityGroup = new SecurityGroup(this, "Ec2SecurityGroup", {
            vpc: vpc,
            description: "Security Group for Ec2 Proxy instance",
            disableInlineRules: true,
            securityGroupName: "Ec2ProxySecurityGroup",
            allowAllOutbound: true,
            allowAllIpv6Outbound: true,
        });

        const instanceRole = new Role(this, 'EC2ProxyRole', {
            roleName: 'EC2ProxyRole',
            description: 'EC2 Proxy Role',
            managedPolicies: [ManagedPolicy.fromAwsManagedPolicyName('AmazonSSMManagedInstanceCore')],
            assumedBy: new ServicePrincipal('ec2.amazonaws.com')
        });

        const profile = new CfnInstanceProfile(this, `InstanceProfile`, {
            roles: [ instanceRole.roleName ],
            instanceProfileName: 'ProxyInstanceProfile'
        });

        const ssmaUserData = UserData.forLinux();
        ssmaUserData.addCommands(
            'yum update -y -q',
            'sudo yum install ec2-instance-connect',
            'sudo systemctl enable amazon-ssm-agent',
            'sudo systemctl start amazon-ssm-agent',
            'sudo yum install squid',
            'sudo service squid restart'
        );

        new CfnInstance(this, 'ProxyServerInstance', {
            subnetId: vpc.privateSubnets[0].subnetId,
            securityGroupIds: [ec2SecurityGroup.securityGroupId],
            iamInstanceProfile: profile.instanceProfileName,
            imageId: new AmazonLinuxImage().getImage(this).imageId,
            instanceType: InstanceType.of(InstanceClass.T2, InstanceSize.MICRO).toString(),
            userData: Fn.base64(ssmaUserData.render()),
            tags: [{key: 'Name', value: 'proxy-server'}],
            });
    }

}