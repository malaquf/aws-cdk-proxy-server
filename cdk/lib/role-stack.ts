import { Construct } from 'constructs';
import { StringParameter } from 'aws-cdk-lib/aws-ssm';
import {Stack, StackProps, Tags} from 'aws-cdk-lib';
import { InterfaceVpcEndpointAwsService, SubnetType, Vpc } from 'aws-cdk-lib/aws-ec2';
import { AccountPrincipal, Effect, PolicyDocument, PolicyStatement, Role } from 'aws-cdk-lib/aws-iam';

export class RoleStack extends Stack {
    constructor(scope: Construct, id: string, props?: StackProps) {
        super(scope, id, props);
        
        const role = new Role(this, 'ProxyUserRole', {
            assumedBy: new AccountPrincipal(this.account),
            roleName: 'proxy-user-role',
            description: 'Role for accessing proxy server',
        });
        role.addToPolicy(new PolicyStatement({
            effect: Effect.ALLOW,
            actions: [
                'ec2:DescribeInstances'
            ],
            resources: ['*']
        }));
        role.addToPolicy(new PolicyStatement({
            effect: Effect.ALLOW,
            actions: [
                'ssm:StartSession',
                'ec2-instance-connect:SendSSHPublicKey'
            ],
            resources: ['arn:aws:ec2:*:*:instance/*'],
            conditions: [
                { 'StringEquals' : { 'aws:ResourceTag/Name': 'proxy-server' }}
            ]
        }));
        role.addToPolicy(new PolicyStatement({
            effect: Effect.ALLOW,
            actions: [
                'ssm:StartSession'
            ],
            resources: ['arn:aws:ssm:*:*:document/AWS-StartSSHSession']
        }));
        role.addToPolicy(new PolicyStatement({
            effect: Effect.ALLOW,
            actions: [
                'ssm:TerminateSession',
                'ssm:ResumeSession'
            ],
            resources: ['arn:aws:ssm:*:*:session/$${aws:username}-*']
        }));
    }
}
