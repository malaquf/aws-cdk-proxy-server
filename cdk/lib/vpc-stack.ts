import { Construct } from 'constructs';
import { StringParameter } from 'aws-cdk-lib/aws-ssm';
import {Stack, StackProps, Tags} from 'aws-cdk-lib';
import { InterfaceVpcEndpointAwsService, SubnetType, Vpc } from 'aws-cdk-lib/aws-ec2';

export class VpcStack extends Stack {
    constructor(scope: Construct, id: string, props?: StackProps) {
        super(scope, id, props);
        
        const vpc = new Vpc(this, 'Vpc', {
            maxAzs: 1, // increase it for resilience
            subnetConfiguration: [
                {
                    cidrMask: 24,
                    name: 'public',
                    subnetType: SubnetType.PUBLIC,
                },
                {
                    cidrMask: 24,
                    name: 'private',
                    subnetType: SubnetType.PRIVATE_WITH_EGRESS,
                }
            ]
        });

        // allows access to EC2 in private subnet through SSM
        vpc.addInterfaceEndpoint('SSMInterfaceVpcEndpoint', {
            service: new InterfaceVpcEndpointAwsService('ssm'),
            subnets: { subnetType: SubnetType.PUBLIC},
            privateDnsEnabled: true,
            open: true
        });
        vpc.addInterfaceEndpoint('Ec2InterfaceVpcEndpoint', {
            service: new InterfaceVpcEndpointAwsService('ec2'),
            subnets: { subnetType: SubnetType.PUBLIC},
            privateDnsEnabled: true,
            open: true
        });
        vpc.addInterfaceEndpoint('SSMMessagesInterfaceVpcEndpoint', {
            service: new InterfaceVpcEndpointAwsService('ssmmessages'),
            subnets: { subnetType: SubnetType.PUBLIC},
            privateDnsEnabled: true,
            open: true
        });
        vpc.addInterfaceEndpoint('EC2MessagesInterfaceVpcEndpoint', {
            service: new InterfaceVpcEndpointAwsService('ec2messages'),
            subnets: { subnetType: SubnetType.PUBLIC},
            privateDnsEnabled: true,
            open: true
        });

        new StringParameter(this, 'VpcId', {
            parameterName: '/VpcId',
            stringValue: vpc.vpcId
        });

    }
}
