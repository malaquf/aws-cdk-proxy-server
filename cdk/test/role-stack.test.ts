import { App } from 'aws-cdk-lib';
import { Template } from 'aws-cdk-lib/assertions';
import { VpcStack } from '../lib/vpc-stack';
import { RoleStack } from '../lib/role-stack';


describe('role stack - snapshot test', () => {

    it('stack is still valid', () => {
        const app = new App();

        const stack = new RoleStack(app, 'MyTestStack');

        expect(Template.fromStack(stack)).toMatchSnapshot();
    });
});
