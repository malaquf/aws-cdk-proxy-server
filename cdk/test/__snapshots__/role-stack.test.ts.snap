// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`role stack - snapshot test stack is still valid 1`] = `
{
  "Parameters": {
    "BootstrapVersion": {
      "Default": "/cdk-bootstrap/hnb659fds/version",
      "Description": "Version of the CDK Bootstrap resources in this environment, automatically retrieved from SSM Parameter Store. [cdk:skip]",
      "Type": "AWS::SSM::Parameter::Value<String>",
    },
  },
  "Resources": {
    "ProxyUserRole8BFC6591": {
      "Properties": {
        "AssumeRolePolicyDocument": {
          "Statement": [
            {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": {
                "AWS": {
                  "Fn::Join": [
                    "",
                    [
                      "arn:",
                      {
                        "Ref": "AWS::Partition",
                      },
                      ":iam::",
                      {
                        "Ref": "AWS::AccountId",
                      },
                      ":root",
                    ],
                  ],
                },
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "Description": "Role for accessing proxy server",
        "RoleName": "proxy-user-role",
      },
      "Type": "AWS::IAM::Role",
    },
    "ProxyUserRoleDefaultPolicy8C9A7745": {
      "Properties": {
        "PolicyDocument": {
          "Statement": [
            {
              "Action": "ec2:DescribeInstances",
              "Effect": "Allow",
              "Resource": "*",
            },
            {
              "Action": [
                "ssm:StartSession",
                "ec2-instance-connect:SendSSHPublicKey",
              ],
              "Condition": {
                "0": {
                  "StringEquals": {
                    "aws:ResourceTag/Name": "proxy-server",
                  },
                },
              },
              "Effect": "Allow",
              "Resource": "arn:aws:ec2:*:*:instance/*",
            },
            {
              "Action": "ssm:StartSession",
              "Effect": "Allow",
              "Resource": "arn:aws:ssm:*:*:document/AWS-StartSSHSession",
            },
            {
              "Action": [
                "ssm:TerminateSession",
                "ssm:ResumeSession",
              ],
              "Effect": "Allow",
              "Resource": "arn:aws:ssm:*:*:session/$\${aws:username}-*",
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "ProxyUserRoleDefaultPolicy8C9A7745",
        "Roles": [
          {
            "Ref": "ProxyUserRole8BFC6591",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
  },
  "Rules": {
    "CheckBootstrapVersion": {
      "Assertions": [
        {
          "Assert": {
            "Fn::Not": [
              {
                "Fn::Contains": [
                  [
                    "1",
                    "2",
                    "3",
                    "4",
                    "5",
                  ],
                  {
                    "Ref": "BootstrapVersion",
                  },
                ],
              },
            ],
          },
          "AssertDescription": "CDK bootstrap stack version 6 required. Please run 'cdk bootstrap' with a recent version of the CDK CLI.",
        },
      ],
    },
  },
}
`;
